### PR TITLE
Add options to pin unread/mentioned rooms to the top of the room list

### PR DESCRIPTION
**Pull request by turt2live**
Fixes https://github.com/vector-im/riot-web/issues/6604
Fixes https://github.com/vector-im/riot-web/issues/732
Fixes https://github.com/vector-im/riot-web/issues/1104
Fixes https://github.com/vector-im/riot-web/issues/7319

### CODE REVIEW

**Comment by turt2live** [comment]
Originally there was a sort in there, however the sort gets complicated due to the tri-state nature: mentioned, notified (unread), and unread. Having the fused list makes it a little easier to understand in my opinion.

**Comment by lukebarnard1** [comment]
>I dislike that this logic for room ordering is now split between here (the component) and RoomListStore.

Two stores updating the same view isn't the worst thing in the world, I don't think. In this case the settings store isn't fluxy but the model still roughly holds.

Having said that, the RoomListStore already imports a lot of stuff and there's a `_recentsComparator` that this could be done in. And there's something to be said about keeping all the sorting logic in the same place.

**Comment by bwindels** [comment]
@turt2live `UserSettings.js` is conflicting with `develop`.

**Comment by turt2live** [comment]
@bwindels thanks. Should be good now

File: src/stores/RoomListStore.js [hunk 1]
```diff
@@ -263,6 +264,30 @@ class RoomListStore extends Store {
...
+             if (mentionsA && mentionsB) return 0;
+             // If neither have mentions, fall through to remaining checks
+        }
+
+        if (pinUnread) {
+            const unreadA = Unread.doesRoomHaveUnreadMessages(roomA);
+            const unreadB = Unread.doesRoomHaveUnreadMessages(roomB);
+            if (unreadA && !unreadB) return -1;
+            if (!unreadA && unreadB) return 1;
+            if (unreadA && unreadB) return 0;
```
Comment (bwindels) [inline src/stores/RoomListStore.js]: `else` to make clear these 3 lines will always return?

File: src/stores/RoomListStore.js [hunk 2]
```diff
@@ -263,6 +264,30 @@ class RoomListStore extends Store {
     }
 
     _recentsComparator(roomA, roomB) {
+        const pinUnread = SettingsStore.getValue("pinUnreadRooms");
```
Comment (bwindels) [inline src/stores/RoomListStore.js]: I wonder if implementing the 3 ways to sort in 3 different methods, and then choosing a comparator based on the UserSettings on [line 224](https://github.com/matrix-org/matrix-react-sdk/pull/1936/files#diff-fbd69d95e670ba89b7a7e1abf60c2480L224) wouldn't look nicer.

File: src/stores/RoomListStore.js (hunk 1 above)
Comment (bwindels) [reply src/stores/RoomListStore.js]: Hmm, we also probably want to keep sorting by timestamp if both rooms don't have unread messages? If we don't the sorting order could be unstable and have rooms jump up and down. Same for mentions.

Comment (bwindels) [reply src/stores/RoomListStore.js]: right! Missed that :man_facepalming:

File: src/stores/RoomListStore.js (hunk 2 above)
Comment (bwindels) [reply src/stores/RoomListStore.js]: Indeed, given the other comment.

**Pull request merged**